contracts/src/SuccinctVApp.sol

```diff
@@ -390,54 +341,77 @@ contract SuccinctVApp is
-            TransactionStatus status = _publicValues.receipts[i].status;
-            if (status == TransactionStatus.None || status == TransactionStatus.Pending) {
-                revert ReceiptStatusInvalid(status);
-            }
+    function _handleOnchainReceipt(Receipt memory _receipt) internal {
+        // Increment the finalized onchain transaction ID.
+        uint64 onchainTxId = ++finalizedOnchainTxId;
 
-            // Ensure that the receipt is consistent with the transaction.
-            Receipts.assertEq(transactions[onchainTxId], _publicValues.receipts[i]);
+        // Ensure that the receipt is the next one to be processed.
+        if (onchainTxId != _receipt.onchainTxId) {
+            revert ReceiptOutOfOrder(onchainTxId, _receipt.onchainTxId);
+        }
 
-            // Update the transaction status.
-            transactions[onchainTxId].status = status;
+        // Ensure that the receipt has of the expected statuses.
+        if (
+            _receipt.status == TransactionStatus.None
+                || _receipt.status == TransactionStatus.Pending
+        ) {
+            revert ReceiptStatusInvalid(_receipt.status);
+        }
 
-            // If the transaction failed, emit the revert event and skip the rest of the loop.
-            TransactionVariant variant = _publicValues.receipts[i].variant;
-            if (status == TransactionStatus.Reverted) {
-                emit TransactionReverted(onchainTxId, variant, _publicValues.receipts[i].action);
-                continue;
-            }
+        // Ensure that the receipt is consistent with the transaction.
+        Receipts.assertEq(transactions[onchainTxId], _receipt);
 
-            // If the transaction completed, run a handler for the transaction.
-            if (variant == TransactionVariant.Deposit) {
-                // No-op.
-            } else if (variant == TransactionVariant.Withdraw) {
-                Withdraw memory withdraw = abi.decode(_publicValues.receipts[i].action, (Withdraw));
-                _processWithdraw(withdraw.account, withdraw.amount);
-            } else if (variant == TransactionVariant.CreateProver) {
-                // No-op.
-            } else {
-                revert TransactionVariantInvalid();
-            }
+        // Update the transaction status.
+        transactions[onchainTxId].status = _receipt.status;
+
+        // If the transaction failed, emit the revert event and skip the rest of the loop.
+        if (_receipt.status == TransactionStatus.Reverted) {
+            emit TransactionReverted(onchainTxId, _receipt.variant, _receipt.action);
+            return;
+        }
 
-            // Emit the completed event.
-            emit TransactionCompleted(onchainTxId, variant, _publicValues.receipts[i].action);
+        // If the transaction completed, run a handler for the transaction.
+        if (_receipt.variant == TransactionVariant.Deposit) {
```

Comment:
cleaner to just amke this if `_receipt.variant != TransactionVariant.Deposit && _receipt.variant != TransactionVariant.CreateProver`, revert